File: navidrome-ios/Store/SyncStore.swift
import Combine
import Foundation
import UIKit

/// Single source of truth for the entire app.
/// Views never touch SyncClient directly — everything goes through the store.
@MainActor
final class SyncStore: ObservableObject {

    // MARK: - Published state

    @Published var nowPlaying: NowPlayingSong?
    @Published var myRole: String = "observer"
    @Published var myClientId: String
    @Published var activeClientId: String?
    @Published var connectedClients: [ConnectedClient] = []
    @Published var isConnected: Bool = false
    @Published var isPlaying: Bool = false
    @Published var position: Double = 0
    @Published var lastSyncTime: Date = .distantPast

    @Published var queue: [NowPlayingSong] = []
    @Published var queueIndex: Int = 0
    @Published var isShuffled: Bool = false
    @Published var repeatMode: RepeatMode = .off

    /// Stores the original (unshuffled) queue order.
    private var originalQueue: [NowPlayingSong] = []

    // MARK: - Private

    private let syncClient = SyncClient()
    let audioPlayer = AudioPlayer()
    private var positionReportingTask: Task<Void, Never>?
    private var cancellables = Set<AnyCancellable>()
    private var interpolationTask: Task<Void, Never>?

    // MARK: - Init

    init() {
        myClientId = AppConfig.clientId
        setupBindings()
    }

    /// Wiring closures in a separate method so `self` is fully initialized
    /// and captured as a let-binding (avoids "captured var self" errors).
    private func setupBindings() {
        // Wire audio player state back into the store
        audioPlayer.$isPlaying
            .receive(on: RunLoop.main)
            .assign(to: &$isPlaying)

        audioPlayer.$currentTime
            .receive(on: RunLoop.main)
            .sink { [weak self] time in
                guard let self, self.myRole == "active" else { return }
                self.position = time
            }
            .store(in: &cancellables)

        audioPlayer.onTrackEnd = { [weak self] in
            Task { @MainActor in self?.handleTrackEnd() }
        }

        // Wire remote command center to store actions
        audioPlayer.onRemotePlay = { [weak self] in Task { @MainActor in self?.play() } }
        audioPlayer.onRemotePause = { [weak self] in Task { @MainActor in self?.pause() } }
        audioPlayer.onRemoteNext = { [weak self] in Task { @MainActor in self?.next() } }
        audioPlayer.onRemotePrev = { [weak self] in Task { @MainActor in self?.prev() } }
        audioPlayer.onRemoteSeek = { [weak self] pos in Task { @MainActor in self?.seek(to: pos) } }

        // WebSocket message handler
        syncClient.onMessage = { [weak self] envelope in
            Task { @MainActor in
                self?.handleMessage(envelope)
            }
        }
        syncClient.onConnected = { [weak self] in
            Task { @MainActor in self?.isConnected = true }
        }
        syncClient.onDisconnected = { [weak self] in
            Task { @MainActor in self?.isConnected = false }
        }


    }

    // MARK: - Connection

    /// Connect to the Go sync service. Only works if a sync service URL is configured
    /// (separate from the Navidrome server URL).
    func connect() {
        guard let base = AppConfig.syncServiceURL, !base.isEmpty else { return }
        syncClient.connect(baseURL: base, clientId: myClientId)
        startPositionReporting()
    }

    func disconnect() {
        syncClient.disconnect()
        stopPositionReporting()
        isConnected = false
    }

    // MARK: - Playback

    func playSong(_ song: NowPlayingSong, isCommand: Bool = false) {
        if myRole != "active" && !isCommand {
            if isConnected {
                let payload = PlaySongPayload(song: song)
                sendMessage(type: .playSong, payload: payload)
            } else {
                becomeActiveLocally()
                queue = [song]
                queueIndex = 0
                loadAndPlay(song)
            }
            return
        }
        queue = [song]
        queueIndex = 0
        loadAndPlay(song)
        if isConnected {
            sendNowPlaying(song)
            sendQueueToHub()
        }
    }

    func playQueue(_ songs: [NowPlayingSong], startIndex: Int, isCommand: Bool = false) {
        guard startIndex < songs.count else { return }
        if myRole != "active" && !isCommand {
            if isConnected {
                let payload = LoadQueuePayload(queue: songs, startIndex: startIndex)
                sendMessage(type: .loadQueue, payload: payload)
            } else {
                becomeActiveLocally()
                queue = songs
                queueIndex = startIndex
                loadAndPlay(songs[startIndex])
            }
            return
        }
        queue = songs
        queueIndex = startIndex
        let song = songs[startIndex]
        loadAndPlay(song)
        if isConnected {
            sendNowPlaying(song)
            sendQueueToHub()
        }
    }

    func play() {
        if myRole != "active" {
            if isConnected { sendMessage(type: .play) }
            return
        }
        audioPlayer.resume()
        if isConnected { sendMessage(type: .play) }
    }

    func pause() {
        if myRole != "active" {
            if isConnected { sendMessage(type: .pause) }
            return
        }
        audioPlayer.pause()
        if isConnected { sendMessage(type: .pause) }
    }

    func seek(to positionSecs: Double) {
        if myRole != "active" {
            if isConnected { sendMessage(type: .seek, payload: SeekPayload(positionSecs: positionSecs)) }
            return
        }
        audioPlayer.seek(to: positionSecs)
        position = positionSecs
        if isConnected { sendMessage(type: .seek, payload: SeekPayload(positionSecs: positionSecs)) }
    }

    func next(isCommand: Bool = false) {
        if myRole != "active" && !isCommand {
            if isConnected { sendMessage(type: .next) }
            return
        }
        
        let nextIndex = queueIndex + 1
        guard nextIndex < queue.count else {
            // If repeat all, wrap around to first track
            if repeatMode == .all && !queue.isEmpty {
                queueIndex = 0
                let song = queue[0]
                loadAndPlay(song)
                if isConnected {
                    sendNowPlaying(song)
                    sendQueueToHub()
                }
            }
            return
        }
        queueIndex = nextIndex
        let song = queue[nextIndex]
        loadAndPlay(song)
        if isConnected {
            sendNowPlaying(song)
            sendQueueToHub()
        }
    }

    func prev(isCommand: Bool = false) {
        if myRole != "active" && !isCommand {
            if isConnected { sendMessage(type: .prev) }
            return
        }
        
        // If more than 3s in, restart current track
        if audioPlayer.currentTime > 3 {
            audioPlayer.seek(to: 0)
            return
        }
        let prevIndex = queueIndex - 1
        guard prevIndex >= 0 else {
            audioPlayer.seek(to: 0)
            return
        }
        queueIndex = prevIndex
        let song = queue[prevIndex]
        loadAndPlay(song)
        if isConnected {
            sendNowPlaying(song)
            sendQueueToHub()
        }
    }

    func claim() {
        if isConnected {
            sendMessage(type: .claim)
        } else {
            becomeActiveLocally()
        }
    }

    func toggleShuffle() {
        isShuffled.toggle()
        if isShuffled {
            originalQueue = queue
            // Shuffle everything after the current index
            let currentSong = queue[queueIndex]
            var remaining = queue
            remaining.remove(at: queueIndex)
            remaining.shuffle()
            queue = [currentSong] + remaining
            queueIndex = 0
        } else {
            // Restore original order, keeping the current song's position
            let currentSong = queue[queueIndex]
            queue = originalQueue
            queueIndex = queue.firstIndex(where: { $0.songId == currentSong.songId }) ?? 0
            originalQueue = []
        }
        if isConnected {
            sendQueueToHub()
            sendPlaybackOptions()
        }
    }

    func toggleRepeat() {
        switch repeatMode {
        case .off: repeatMode = .all
        case .all: repeatMode = .one
        case .one: repeatMode = .off
        }
        if isConnected { sendPlaybackOptions() }
    }

    func clearQueue() {
        // Keep only the currently playing song
        if queueIndex < queue.count {
            let currentSong = queue[queueIndex]
            queue = [currentSong]
            queueIndex = 0
        }
        if isConnected { sendQueueToHub() }
    }

    /// Called when a track finishes playing.
    private func handleTrackEnd() {
        switch repeatMode {
        case .one:
            // Replay the same track
            if let song = nowPlaying {
                loadAndPlay(NowPlayingSong(
                    songId: song.songId, title: song.title, artist: song.artist,
                    album: song.album, coverArtId: song.coverArtId,
                    durationSecs: song.durationSecs, positionSecs: 0
                ))
            }
        case .all, .off:
            next()
        }
    }

    /// Become the active client locally without waiting for sync service confirmation.
    private func becomeActiveLocally() {
        myRole = "active"
        stopInterpolation()
        startPositionReporting()
    }

    // MARK: - Private helpers

    private func loadAndPlay(_ song: NowPlayingSong) {
        nowPlaying = song
        guard let url = NavidromeClient.shared.streamURL(songId: song.songId) else { return }
        audioPlayer.play(url: url, position: song.positionSecs)
        updateLockScreen(song: song)
    }

    private func updateLockScreen(song: NowPlayingSong) {
        audioPlayer.updateNowPlayingInfo(
            title: song.title,
            artist: song.artist,
            album: song.album,
            duration: Double(song.durationSecs),
            position: song.positionSecs
        )
    }

    // MARK: - WebSocket messaging

    private func sendMessage(type: MessageType, payload: (any Encodable)? = nil) {
        syncClient.send(type: type, payload: payload)
    }

    private func sendNowPlaying(_ song: NowPlayingSong) {
        let payload = NowPlayingPayload(
            songId: song.songId,
            title: song.title,
            artist: song.artist,
            album: song.album,
            coverArtId: song.coverArtId,
            durationSecs: song.durationSecs,
            positionSecs: song.positionSecs
        )
        sendMessage(type: .nowPlaying, payload: payload)
    }

    private func sendQueueToHub() {
        let items = queue.map {
            QueueItemPayload(
                songId: $0.songId,
                title: $0.title,
                artist: $0.artist,
                album: $0.album,
                coverArtId: $0.coverArtId,
                durationSecs: $0.durationSecs
            )
        }
        sendMessage(type: .setQueue, payload: SetQueuePayload(queue: items, queueIndex: queueIndex))
    }

    private func sendPlaybackOptions() {
        let modeString: String
        switch repeatMode {
        case .off: modeString = "off"
        case .all: modeString = "all"
        case .one: modeString = "one"
        }
        sendMessage(type: .setPlaybackOptions, payload: PlaybackOptionsPayload(shuffle: isShuffled, repeatMode: modeString))
    }

    // MARK: - Position reporting (active only, ~1s)

    private func startPositionReporting() {
        stopPositionReporting()
        positionReportingTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                guard myRole == "active", isPlaying else { continue }
                sendMessage(
                    type: .positionUpdate,
                    payload: PositionUpdatePayload(positionSecs: audioPlayer.currentTime)
                )
            }
        }
    }

    private func stopPositionReporting() {
        positionReportingTask?.cancel()
        positionReportingTask = nil
    }

    // MARK: - Observer interpolation

    private func startInterpolation() {
        stopInterpolation()
        interpolationTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                guard myRole == "observer", nowPlaying != nil else { continue }
                position += 1
            }
        }
    }

    private func stopInterpolation() {
        interpolationTask?.cancel()
        interpolationTask = nil
    }

    // MARK: - Incoming message routing

    private func handleMessage(_ envelope: SyncEnvelope) {
        switch envelope.type {
        case .stateSync:
            guard let payload = envelope.payload?.decode(StateSyncPayload.self) else { return }
            handleStateSync(payload)
        case .command:
            guard let payload = envelope.payload?.decode(CommandPayload.self) else { return }
            handleCommand(payload)
        case .error:
            if let payload = envelope.payload?.decode(ErrorPayload.self) {
                print("[sync] error \(payload.code): \(payload.message)")
            }
        default:
            break
        }
    }

    private func handleStateSync(_ payload: StateSyncPayload) {
        let previousRole = myRole
        activeClientId = payload.activeClientId
        connectedClients = payload.clients
        lastSyncTime = Date()

        // Determine our role from the clients list
        if let me = payload.clients.first(where: { $0.clientId == myClientId }) {
            myRole = me.role
        }

        // Active clients own the state. We don't overwrite local state with echoes
        // of our own messages, as it causes race conditions (e.g. going back and forth).
        let justBecameActive = (previousRole != "active" && myRole == "active")
        
        if myRole == "observer" || justBecameActive {
            // Accept queue from server
            if let serverQueue = payload.queue {
                queue = serverQueue.map { $0.toNowPlayingSong() }
                queueIndex = payload.queueIndex ?? 0
            }

            // Accept shuffle & repeat from server
            if let shuffle = payload.shuffle {
                isShuffled = shuffle
            }
            if let rm = payload.repeatMode {
                switch rm {
                case "all": repeatMode = .all
                case "one": repeatMode = .one
                default: repeatMode = .off
                }
            }
        }

        // Update now playing from server state
        if let song = payload.song {
            if myRole == "observer" || justBecameActive {
                nowPlaying = song
                if myRole == "observer" {
                    position = song.positionSecs
                    startInterpolation()
                }
            }
            
            // Auto-play when we just became active and there's a song
            if justBecameActive {
                loadAndPlay(song)
            }
        } else {
            if myRole == "observer" || justBecameActive {
                nowPlaying = nil
                stopInterpolation()
            }
        }
    }

    private func handleRoleChange(_ payload: RoleChangePayload) {
        guard payload.clientId == myClientId else { return }
        myRole = payload.role
        if payload.role == "observer" {
            stopPositionReporting()
            startInterpolation()
        } else {
            stopInterpolation()
            startPositionReporting()
        }
    }

    private func handleCommand(_ payload: CommandPayload) {
        switch payload.action {
        case "STOP":
            audioPlayer.stop()
        case "PLAY":
            audioPlayer.resume()
        case "PAUSE":
            audioPlayer.pause()
        case "SEEK":
            if let pos = payload.positionSecs {
                audioPlayer.seek(to: pos)
            }
        case "NEXT":
            next(isCommand: true)
        case "PREV":
            prev(isCommand: true)
        case "PLAY_SONG":
            if let songPayload = payload.song {
                let song = NowPlayingSong(
                    songId: songPayload.songId,
                    title: songPayload.title,
                    artist: songPayload.artist,
                    album: songPayload.album,
                    coverArtId: songPayload.coverArtId,
                    durationSecs: songPayload.durationSecs,
                    positionSecs: songPayload.positionSecs ?? 0.0
                )
                playSong(song, isCommand: true)
            }
        case "LOAD_QUEUE":
            if let q = payload.queue, let startIndex = payload.startIndex {
                let songs = q.map { $0.toNowPlayingSong() }
                playQueue(songs, startIndex: startIndex, isCommand: true)
            }
        default:
            break
        }
    }
}

// MARK: - Repeat mode

enum RepeatMode {
    case off, all, one
}

File: web/src/store/syncStore.ts
import { create } from 'zustand';
import { streamUrl } from '../api/navidrome';

export interface NowPlayingSong {
  songId: string;
  title: string;
  artist: string;
  album: string;
  coverArtId: string;
  durationSecs: number;
  positionSecs: number;
  isPlaying?: boolean;
}

export interface ConnectedClient {
  clientId: string;
  clientType: 'web' | 'ios';
  role: 'active' | 'observer';
}

// Persistent audio element — lives for the lifetime of the app.
const audio = new Audio();

export type RepeatMode = 'off' | 'all' | 'one';

interface SyncState {
  nowPlaying: NowPlayingSong | null;
  myClientId: string;
  myRole: 'active' | 'observer';
  activeClientId: string | null;
  connectedClients: ConnectedClient[];
  isConnected: boolean;
  lastSyncTime: number;

  // Playback state driven by the persistent <audio> element
  isPlaying: boolean;
  position: number;

  // Queue
  queue: NowPlayingSong[];
  queueIndex: number;

  // Shuffle & Repeat
  shuffle: boolean;
  repeatMode: RepeatMode;

  // WebSocket send function, set by useWebSocket
  sendMessage: ((type: string, payload?: Record<string, unknown>) => void) | null;
  setSendMessage: (fn: (type: string, payload?: Record<string, unknown>) => void) => void;

  setConnected: (connected: boolean) => void;
  setMyClientId: (id: string) => void;
  handleStateSync: (payload: {
    activeClientId: string | null;
    song: NowPlayingSong | null;
    clients: ConnectedClient[];
    queue?: NowPlayingSong[];
    queueIndex?: number;
    shuffle?: boolean;
    repeatMode?: RepeatMode;
  }) => void;
  handleRoleChange: (payload: { clientId: string; role: 'active' | 'observer' }) => void;
  handleCommand: (payload: { action: string; positionSecs?: number; song?: NowPlayingSong; queue?: NowPlayingSong[]; startIndex?: number; queueIndex?: number }) => void;
  handleError: (payload: { code: string; message: string }) => void;

  /** Get the underlying HTMLAudioElement (for direct binding in AudioManager). */
  getAudio: () => HTMLAudioElement;

  /** Load a song into the persistent audio and start playback. */
  playSong: (song: NowPlayingSong, isCommand?: boolean) => void;
  /** Set the queue and start playing a specific index. */
  playQueue: (queue: NowPlayingSong[], startIndex: number, isCommand?: boolean) => void;
  play: () => void;
  pause: () => void;
  seek: (positionSecs: number) => void;
  next: (isCommand?: boolean) => void;
  prev: (isCommand?: boolean) => void;

  toggleShuffle: () => void;
  cycleRepeatMode: () => void;
  removeFromQueue: (index: number) => void;
  playQueueIndex: (index: number, isCommand?: boolean) => void;
  clearQueue: () => void;
  showQueue: boolean;
  setShowQueue: (show: boolean) => void;

  // Playlist invalidation
  lastPlaylistInvalidation: { playlistId: string; action: string } | null;
  notifyPlaylistChanged: (playlistId: string, action: string) => void;

  claim: () => void;
  sendCommand: (type: 'PLAY' | 'PAUSE' | 'NEXT' | 'PREV' | 'SEEK' | 'PLAY_SONG' | 'LOAD_QUEUE' | 'PLAY_INDEX', payload?: Record<string, unknown>) => void;
  sendNowPlaying: (song: NowPlayingSong) => void;
  sendPositionUpdate: (positionSecs: number) => void;
  sendQueue: (queue: NowPlayingSong[], queueIndex: number) => void;
  sendPlaybackOptions: () => void;
}

function generateClientId(): string {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  // Fallback for non-secure contexts (plain HTTP)
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export const useSyncStore = create<SyncState>((set, get) => ({
  nowPlaying: null,
  myClientId: generateClientId(),
  myRole: 'observer',
  activeClientId: null,
  connectedClients: [],
  isConnected: false,
  lastSyncTime: 0,
  isPlaying: false,
  position: 0,
  queue: [],
  queueIndex: 0,
  shuffle: false,
  repeatMode: 'off' as RepeatMode,
  showQueue: false,
  sendMessage: null,
  lastPlaylistInvalidation: null,

  setSendMessage: (fn) => set({ sendMessage: fn }),

  setConnected: (connected) => set({ isConnected: connected }),

  setMyClientId: (id) => set({ myClientId: id }),

  handleStateSync: (payload) => {
    const { myClientId, myRole: previousRole } = get();
    const myClient = payload.clients.find((c) => c.clientId === myClientId);
    const newRole = myClient?.role ?? 'observer';
    const justBecameActive = previousRole !== 'active' && newRole === 'active';
    
    const updates: Partial<SyncState> = {
      activeClientId: payload.activeClientId,
      connectedClients: payload.clients,
      myRole: newRole,
      lastSyncTime: Date.now(),
    };
    
    // Accept queue and state from server if we're an observer (or just became active via claim)
    if (newRole === 'observer' || justBecameActive) {
      updates.nowPlaying = payload.song;
      // Reflect the active client's actual play/pause state for observers.
      if (newRole === 'observer') {
        updates.isPlaying = payload.song?.isPlaying ?? false;
      }
      if (payload.queue) {
        updates.queue = payload.queue.map((q) => ({ ...q, positionSecs: 0 }));
        updates.queueIndex = payload.queueIndex ?? 0;
      }
      if (payload.shuffle != null) {
        updates.shuffle = payload.shuffle;
      }
      if (payload.repeatMode != null) {
        updates.repeatMode = payload.repeatMode;
      }
    }
    
    set(updates);

    // When we just became active and there's a song, load it.
    // Only auto-play if the hub says the song was actually playing.
    if (justBecameActive && payload.song) {
      const song = payload.song;
      const shouldPlay = song.isPlaying;
      if (!audio.paused && audio.src.includes(song.songId)) {
        if (Math.abs(audio.currentTime - song.positionSecs) > 2) {
          audio.currentTime = song.positionSecs;
        }
        if (!shouldPlay) {
          audio.pause();
        }
      } else {
        audio.src = streamUrl(song.songId);
        if (song.positionSecs > 0) {
          audio.currentTime = song.positionSecs;
        }
        if (shouldPlay) {
          audio.play().catch(() => {});
        }
      }
      set({ isPlaying: shouldPlay });
    }
  },

  handleRoleChange: (payload) => {
    const { myClientId } = get();
    if (payload.clientId === myClientId) {
      set({ myRole: payload.role });
    }
  },

  handleCommand: (payload) => {
    const store = get();
    switch (payload.action) {
      case 'STOP':
        audio.pause();
        audio.src = '';
        break;
      case 'PLAY':
        audio.play().catch(() => {});
        break;
      case 'PAUSE':
        audio.pause();
        break;
      case 'SEEK':
        if (payload.positionSecs != null) {
          audio.currentTime = payload.positionSecs;
        }
        break;
      case 'NEXT':
        store.next(true); // pass true to indicate it's an incoming command execution
        break;
      case 'PREV':
        store.prev(true);
        break;
      case 'PLAY_SONG':
        if (payload.song) {
          store.playSong(payload.song, true);
        }
        break;
      case 'LOAD_QUEUE':
        if (payload.queue && payload.startIndex !== undefined) {
          store.playQueue(payload.queue, payload.startIndex, true);
        }
        break;
      case 'PLAY_INDEX':
        if (payload.queueIndex !== undefined) {
          store.playQueueIndex(payload.queueIndex, true);
        }
        break;
    }
  },

  handleError: (payload) => {
    console.error(`[sync] ${payload.code}: ${payload.message}`);
  },

  getAudio: () => audio,

  playSong: (song, isCommand = false) => {
    const { myRole, sendNowPlaying, sendQueue } = get();
    if (myRole !== 'active' && !isCommand) {
      get().sendCommand('PLAY_SONG', { song });
      return;
    }
    set({ queue: [song], queueIndex: 0, nowPlaying: song });
    audio.src = streamUrl(song.songId);
    audio.play().catch(() => {});
    if (song.positionSecs > 0) {
      audio.currentTime = song.positionSecs;
    }
    sendQueue([song], 0);
    sendNowPlaying(song);
  },

  playQueue: (queue, startIndex, isCommand = false) => {
    const song = queue[startIndex];
    if (!song) return;
    const { myRole, sendNowPlaying, sendQueue } = get();
    if (myRole !== 'active' && !isCommand) {
      get().sendCommand('LOAD_QUEUE', { queue, startIndex });
      return;
    }
    set({ queue, queueIndex: startIndex, nowPlaying: song });
    audio.src = streamUrl(song.songId);
    audio.play().catch(() => {});
    sendQueue(queue, startIndex);
    sendNowPlaying(song);
  },

  play: () => {
    if (get().myRole !== 'active') {
      get().sendCommand('PLAY');
      return;
    }
    audio.play().catch(() => {});
    get().sendCommand('PLAY');
  },

  pause: () => {
    if (get().myRole !== 'active') {
      get().sendCommand('PAUSE');
      return;
    }
    audio.pause();
    get().sendCommand('PAUSE');
  },

  seek: (positionSecs) => {
    if (get().myRole !== 'active') {
      get().sendCommand('SEEK', { positionSecs });
      return;
    }
    audio.currentTime = positionSecs;
    get().sendCommand('SEEK', { positionSecs });
  },

  next: (isCommand = false) => {
    const { queue, queueIndex, sendNowPlaying, sendQueue, myRole, shuffle, repeatMode } = get();
    if (myRole !== 'active' && !isCommand) {
      get().sendCommand('NEXT');
      return;
    }
    if (queue.length === 0) return;

    // Repeat-one: replay the current track
    if (repeatMode === 'one') {
      audio.currentTime = 0;
      audio.play().catch(() => {});
      return;
    }

    let nextIndex: number;
    if (shuffle) {
      // Pick a random index different from current (if possible)
      if (queue.length === 1) {
        nextIndex = 0;
      } else {
        do {
          nextIndex = Math.floor(Math.random() * queue.length);
        } while (nextIndex === queueIndex);
      }
    } else {
      nextIndex = queueIndex + 1;
      if (nextIndex >= queue.length) {
        if (repeatMode === 'all') {
          nextIndex = 0;
        } else {
          return; // end of queue, no repeat
        }
      }
    }

    const song = queue[nextIndex];
    set({ queueIndex: nextIndex, nowPlaying: song });
    audio.src = streamUrl(song.songId);
    audio.play().catch(() => {});
    sendNowPlaying(song);
    sendQueue(queue, nextIndex);
  },

  prev: (isCommand = false) => {
    const { queue, queueIndex, sendNowPlaying, sendQueue, myRole, repeatMode } = get();
    if (myRole !== 'active' && !isCommand) {
      get().sendCommand('PREV');
      return;
    }
    // If more than 3s into the song, restart it; otherwise go to previous
    if (audio.currentTime > 3) {
      audio.currentTime = 0;
      return;
    }
    let prevIndex = queueIndex - 1;
    if (prevIndex < 0) {
      if (repeatMode === 'all') {
        prevIndex = queue.length - 1;
      } else {
        audio.currentTime = 0;
        return;
      }
    }
    const song = queue[prevIndex];
    set({ queueIndex: prevIndex, nowPlaying: song });
    audio.src = streamUrl(song.songId);
    audio.play().catch(() => {});
    sendNowPlaying(song);
    sendQueue(queue, prevIndex);
  },

  toggleShuffle: () => {
    set((s) => ({ shuffle: !s.shuffle }));
    // Defer so the state is updated before we read it
    setTimeout(() => get().sendPlaybackOptions(), 0);
  },

  cycleRepeatMode: () => {
    set((s) => {
      const modes: RepeatMode[] = ['off', 'all', 'one'];
      const idx = modes.indexOf(s.repeatMode);
      return { repeatMode: modes[(idx + 1) % modes.length] };
    });
    setTimeout(() => get().sendPlaybackOptions(), 0);
  },

  removeFromQueue: (index: number) => {
    const { queue, queueIndex, sendQueue } = get();
    if (index < 0 || index >= queue.length) return;
    const newQueue = [...queue];
    newQueue.splice(index, 1);
    let newIndex = queueIndex;
    if (index < queueIndex) {
      newIndex = queueIndex - 1;
    } else if (index === queueIndex && newIndex >= newQueue.length) {
      newIndex = Math.max(0, newQueue.length - 1);
    }
    set({ queue: newQueue, queueIndex: newIndex });
    sendQueue(newQueue, newIndex);
  },

  playQueueIndex: (index: number, isCommand = false) => {
    const { queue, sendNowPlaying, sendQueue, myRole } = get();
    if (index < 0 || index >= queue.length) return;
    const song = queue[index];
    if (myRole !== 'active' && !isCommand) {
      get().sendCommand('PLAY_INDEX', { queueIndex: index });
      return;
    }
    set({ queueIndex: index, nowPlaying: song });
    audio.src = streamUrl(song.songId);
    audio.play().catch(() => {});
    sendNowPlaying(song);
    sendQueue(queue, index);
  },

  clearQueue: () => {
    const { queueIndex, queue, sendQueue } = get();
    // Keep only the currently playing song
    const current = queue[queueIndex];
    if (current) {
      set({ queue: [current], queueIndex: 0 });
      sendQueue([current], 0);
    } else {
      set({ queue: [], queueIndex: 0 });
      sendQueue([], 0);
    }
  },

  setShowQueue: (show: boolean) => set({ showQueue: show }),

  notifyPlaylistChanged: (playlistId, action) => {
    get().sendMessage?.('PLAYLIST_CHANGED', { playlistId, action });
    set({ lastPlaylistInvalidation: { playlistId, action } });
  },

  claim: () => {
    get().sendMessage?.('CLAIM');
    // Start audio immediately inside the user gesture context so the browser
    // autoplay policy doesn't block the subsequent WebSocket-callback play().
    const { nowPlaying } = get();
    if (nowPlaying) {
      audio.src = streamUrl(nowPlaying.songId);
      if (nowPlaying.positionSecs > 0) {
        audio.currentTime = nowPlaying.positionSecs;
      }
      audio.play().catch(() => {});
    }
  },

  sendCommand: (type, payload) => {
    get().sendMessage?.(type, payload);
  },

  sendNowPlaying: (song) => {
    get().sendMessage?.('NOW_PLAYING', song as unknown as Record<string, unknown>);
  },

  sendPositionUpdate: (positionSecs) => {
    get().sendMessage?.('POSITION_UPDATE', { positionSecs });
  },

  sendQueue: (queue, queueIndex) => {
    const items = queue.map((s) => ({
      songId: s.songId,
      title: s.title,
      artist: s.artist,
      album: s.album,
      coverArtId: s.coverArtId,
      durationSecs: s.durationSecs,
    }));
    get().sendMessage?.('SET_QUEUE', { queue: items, queueIndex } as unknown as Record<string, unknown>);
  },

  sendPlaybackOptions: () => {
    const { shuffle, repeatMode } = get();
    get().sendMessage?.('SET_PLAYBACK_OPTIONS', { shuffle, repeatMode });
  },
}));

// Wire audio element events back into the store.
audio.addEventListener('play', () => useSyncStore.setState({ isPlaying: true }));
audio.addEventListener('pause', () => useSyncStore.setState({ isPlaying: false }));
audio.addEventListener('timeupdate', () => {
  useSyncStore.setState({ position: audio.currentTime });
});
audio.addEventListener('ended', () => {
  useSyncStore.getState().next();
});
